com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -504,6 +509,196 @@ internal void InvokeBehaviourOnGainedOwnership()
             }
         }
 
+        private bool m_IsReparented;
+        private ulong? m_LatestParent;
+        private Transform m_CachedParent;
+
+        internal void InvokeBehaviourOnNetworkObjectParentChanged(NetworkObject parentNetworkObject)
+        {
+            for (int i = 0; i < ChildNetworkBehaviours.Count; i++)
+            {
+                ChildNetworkBehaviours[i].OnNetworkObjectParentChanged(parentNetworkObject);
+            }
+        }
+
+        internal void SetCachedParent(Transform parentTransform)
+        {
+            m_CachedParent = parentTransform;
+        }
+
+        private void OnTransformParentChanged()
+        {
+            if (transform.parent == m_CachedParent)
+            {
+                return;
+            }
+
+            if (NetworkManager == null || !NetworkManager.IsListening)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotListeningException($"{nameof(NetworkManager)} is not listening, start a server or host before reparenting"));
+                return;
+            }
+
+            if (!NetworkManager.IsServer)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotServerException($"Only the server can reparent {nameof(NetworkObject)}s"));
+                return;
+            }
+
+            if (!IsSpawned)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new SpawnStateException($"{nameof(NetworkObject)} can only be reparented after being spawned"));
+                return;
+            }
+
+            var parentTransform = transform.parent;
+            if (parentTransform != null)
+            {
+                var parentObject = transform.parent.GetComponent<NetworkObject>();
+                if (parentObject != null)
+                {
+                    m_LatestParent = parentObject.NetworkObjectId;
+                }
+                else
+                {
+                    if (NetworkLog.CurrentLogLevel <= LogLevel.Error)
+                    {
+                        NetworkLog.LogError($"Invalid parenting, {nameof(NetworkObject)} moved under a non-{nameof(NetworkObject)} parent, will fix and move to the scene root");
+                    }
+
+                    m_LatestParent = null;
+                }
+            }
+            else
+            {
+                m_LatestParent = null;
+            }
+
+            using (var buffer = PooledNetworkBuffer.Get())
+            {
+                using (var writer = PooledNetworkWriter.Get(buffer))
+                {
+                    writer.WriteUInt64Packed(NetworkObjectId);
+                    WriteNetworkParenting(writer, m_IsReparented, m_LatestParent);
```

Comment:
If we reparent and move (NetworkTransform) in the same frame, do we have a guarantee the reparent and move will happen at the same time and in the right order client side?

Reply:
we (still) have no guarantees or message delivery orders. well, we kind of have but it is very implicit and opaque. if you were to use the same reliable ordered fragmented channel (or whatever it's called), you could guarantee that they'd be in order — but, that's very subtle and not reliable IMHO.
so, the answer is yes but actually no. also `NetworkObject` parenting doesn't necessarily look into `NetworkTransform` compatibility since it's lower-level than `NetworkTransform`. if things break and/or don't look good, we might even consider reworking `NetworkTransform` after `NetworkObject` changes based on new rules and nuances.

Reply:
there's plans to have grouping for netvars, enabling a guarantee these netvars will be sent together. @jeffreyrainy for visibility.
using netvars would help in this case.

Reply:
I'm not a part of NetVars convos nor snapshotting convos. I'm living in my own silo here and delivering what I'm supposed to deliver.

Reply:
So assuming we have that grouping, that'd make it great to use with this no?

Reply:
again, I have no idea what you folks are talking about or planning. however, I'd still be opinionated about avoiding coupling as much as possible. I'm also hesitant to suggest any architectural or technical changes either in netvars/snapshots space or MLAPI in general. if you have anything fancy in your mind, I'm all ears.